code/modules/jobs/job_exp.dm

```diff
@@ -0,0 +1,203 @@
+	var/list/play_records = prefs.exp
+	if(!play_records.len)
+		update_exp_client(0, 0)
+		play_records = prefs.exp
+		if(!play_records.len)
+			return "[key] has no records."
+	var/return_text = "<UL>"
+	var/list/exp_data = list()
+	for(var/category in GLOB.exp_jobsmap)
+		if(text2num(play_records[category]))
+			exp_data[category] = text2num(play_records[category])
+		else
+			exp_data[category] = 0
+	for(var/dep in exp_data)
+		if(exp_data[dep] > 0)
+			if(dep == EXP_TYPE_EXEMPT)
+				return_text += "<LI>Exempt (all jobs auto-unlocked)</LI>"
+			else if(exp_data[EXP_TYPE_LIVING] > 0)
+				var/my_pc = num2text(round(exp_data[dep]/exp_data[EXP_TYPE_LIVING]*100))
+				return_text += "<LI>[dep] [get_exp_format(exp_data[dep])] ([my_pc]%)</LI>"
+			else
+				return_text += "<LI>[dep] [get_exp_format(exp_data[dep])] </LI>"
+	if(config.use_exp_restrictions_admin_bypass && check_rights(R_ADMIN, 0, mob))
+		return_text += "<LI>Admin (all jobs auto-unlocked)</LI>"
+	return_text += "</UL>"
+	var/list/jobs_locked = list()
+	var/list/jobs_unlocked = list()
+	for(var/datum/job/job in SSjob.occupations)
+		if(job.exp_requirements && job.exp_type)
+			if(!job_is_xp_locked(job.title))
+				continue
+			else if(!job.available_in_playtime(mob.client))
+				jobs_unlocked += job.title
+			else
+				var/xp_req = job.get_exp_req_amount()
+				jobs_locked += "[job.title] [get_exp_format(text2num(play_records[job.get_exp_req_type()]))] / [get_exp_format(xp_req)] as [job.get_exp_req_type()])"
+	if(jobs_unlocked.len)
+		return_text += "<BR><BR>Jobs Unlocked:<UL><LI>"
+		return_text += jobs_unlocked.Join("</LI><LI>")
+		return_text += "</LI></UL>"
+	if(jobs_locked.len)
+		return_text += "<BR><BR>Jobs Not Unlocked:<UL><LI>"
+		return_text += jobs_locked.Join("</LI><LI>")
+		return_text += "</LI></UL>"
+	return return_text
+
+
+/client/proc/get_exp_living()
+	var/exp_living = text2num(prefs.exp[EXP_TYPE_LIVING])
+	return get_exp_format(exp_living)
+
+/proc/get_exp_format(var/expnum)
+	if(expnum > 60)
+		return num2text(round(expnum / 60)) + "h"
+	else if(expnum > 0)
+		return num2text(expnum) + "m"
+	else
+		return "0h"
+
+/proc/update_exp(var/mins, var/ann = 0)
+	if(!SSdbcore.Connect())
+		return -1
+	spawn(0)
+		for(var/client/L in GLOB.clients)
+			if(L.inactivity >= 6000)
+				continue
+			spawn(0)
+				L.update_exp_client(mins, ann)
+			sleep(10)
+
+/client/proc/update_exp_client(var/minutes, var/announce_changes = 0)
+	if(!src ||!ckey)
+		return
+	var/datum/DBQuery/exp_read = SSdbcore.NewQuery("SELECT living, crew, command, engineering, medical, science, supply, security, silicon, service, special, ghost, exempt FROM [format_table_name("role_time")] WHERE ckey='[ckey]'")
+	if(!exp_read.Execute())
+		var/err = exp_read.ErrorMsg()
+		log_game("SQL ERROR during exp_update_client read. Error : \[[err]\]\n")
+		message_admins("SQL ERROR during exp_update_client read. Error : \[[err]\]\n")
+		return
+	var/list/read_records = list()
+	while(exp_read.NextRow())
```

Comment:
There's no failure condition for what happens if i.e. the user doesn't have a record since this is the first time the proc has been run for them. You should add one otherwise the proc will just go on to probably cause trouble with empty values.